internal/functions/compare_versions.go

```diff
@@ -0,0 +1,163 @@
+var _ function.Function = CompareVersionsFunction{}
+
+func NewCompareVersionsFunction() function.Function {
+	return &CompareVersionsFunction{}
+}
+
+type CompareVersionsFunction struct{}
+
+func (f CompareVersionsFunction) Metadata(ctx context.Context, req function.MetadataRequest, resp *function.MetadataResponse) {
+	resp.Name = "compare_versions"
+}
+
+func (f CompareVersionsFunction) Definition(ctx context.Context, req function.DefinitionRequest, resp *function.DefinitionResponse) {
+	resp.Definition = function.Definition{
+		Summary:     "Compare two version strings",
+		Description: "Given two version strings and a comparison operator, returns a boolean result of the comparison.",
+
+		Parameters: []function.Parameter{
+			function.StringParameter{
+				Name:        "version1",
+				Description: "First version string",
+			},
+			function.StringParameter{
+				Name:        "operator",
+				Description: "Comparison operator",
+			},
+			function.StringParameter{
+				Name:        "version2",
+				Description: "Second version string",
+			},
+		},
+		Return: function.BoolReturn{},
+	}
+}
+
+func (f CompareVersionsFunction) Run(ctx context.Context, req function.RunRequest, resp *function.RunResponse) {
+	var version1, version2, operator string
+
+	resp.Error = function.ConcatFuncErrors(req.Arguments.Get(ctx, &version1, &operator, &version2))
+	if resp.Error != nil {
+		return
+	}
+
+	versionResult1 := ParseVersion(version1)
+
+	if versionResult1.Error == "unknown" {
+		resp.Error = function.NewFuncError(fmt.Sprintf("Invalid APIC version provided: %s", version1))
+		return
+	}
+
+	versionResult2 := ParseVersion(version2)
+
+	if versionResult2.Error == "unknown" {
+		resp.Error = function.NewFuncError(fmt.Sprintf("Invalid APIC version provided: %s", version2))
+		return
+	}
+
+	v1 := *versionResult1.Version
+
+	v2 := *versionResult2.Version
+
+	var result bool
+	switch operator {
+	case "==":
+		result = IsVersionEqual(v1, v2)
+	case "!=":
+		result = IsVersionNotEqual(v1, v2)
+	case ">":
+		result = IsVersionGreater(v1, v2)
+	case "<":
+		result = IsVersionLesser(v1, v2)
+	case ">=":
+		result = IsVersionGreaterOrEqual(v1, v2)
+	case "<=":
+		result = IsVersionLessOrEqual(v1, v2)
+	default:
+		resp.Error = function.NewFuncError(fmt.Sprintf("Invalid operator: %s", operator))
+		return
+	}
+
+	resp.Error = function.ConcatFuncErrors(resp.Result.Set(ctx, result))
+}
+
+type Version struct {
+	Major int
+	Minor int
+	Patch int
+	Tag   int
+}
+
+type VersionResult struct {
+	Version *Version
+	Error   string
+}
+
+func ParseVersion(rawVersion string) VersionResult {
+	versionRegex := regexp.MustCompile(`(\d+)\.(\d+)\((\d+)([a-z])\)`)
+	matches := versionRegex.FindStringSubmatch(rawVersion)
+	if matches == nil {
+		return VersionResult{Error: "unknown"}
+	}
+	major, _ := strconv.Atoi(matches[1])
+	minor, _ := strconv.Atoi(matches[2])
+	patch, _ := strconv.Atoi(matches[3])
+	tag := int(matches[4][0])
```

Comment:
Why use something different here? Do you need the [0] as your capture group in regex is already limited to 1 char

Reply:
I get the ASCII value of the string by using int(matches[4][0]) which converts it to uint8 and then I cast it to an int